src/auth/src/credentials/mds_credential.rs

```diff
@@ -0,0 +1,332 @@
+pub(crate) struct MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    token_provider: T,
+}
+
+#[async_trait::async_trait]
+impl<T> Credential for MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    async fn get_token(&mut self) -> Result<Token> {
+        self.token_provider.get_token().await
+    }
+
+    async fn get_headers(&mut self) -> Result<Vec<(HeaderName, HeaderValue)>> {
+        let token = self.get_token().await?;
+        let mut value = HeaderValue::from_str(&format!("{} {}", token.token_type, token.token))
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+        value.set_sensitive(true);
+        Ok(vec![(AUTHORIZATION, value)])
+    }
+
+    async fn get_universe_domain(&mut self) -> Option<String> {
+        Some("googleapis.com".to_string())
+    }
+}
+
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+#[derive(serde::Deserialize, serde::Serialize, PartialEq, Debug, Clone)]
+struct ServiceAccountInfo {
+    email: String,
+    scopes: Option<Vec<String>>,
+    aliases: Option<Vec<String>>,
+}
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+struct MDSAccessTokenProvider {
+    token_endpoint: String,
+}
+
+#[allow(dead_code)]
+impl MDSAccessTokenProvider {
+    async fn get_service_account_info(
+        request: &Client,
+        metadata_service_endpoint: String,
+        email: Option<String>,
+    ) -> Result<ServiceAccountInfo> {
+        let email: String = email.unwrap_or("default".to_string());
+        let path: String = format!(
+            "{}/instance/service-accounts/{}/",
+            metadata_service_endpoint, email
+        );
+        let params = HashMap::from([("recursive", "true")]);
+
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            METADATA_FLAVOR,
+            HeaderValue::from_static(METADATA_FLAVOR_VALUE),
+        );
+
+        let url = reqwest::Url::parse_with_params(path.as_str(), params.iter())
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+
+        let response = request
+            .get(url.clone())
+            .headers(headers)
+            .send()
+            .await
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+
+        response
+            .json::<ServiceAccountInfo>()
+            .await
+            .map_err(|e| CredentialError::new(false, e.into()))
+    }
+}
+
+#[async_trait]
+#[allow(dead_code)]
+impl TokenProvider for MDSAccessTokenProvider {
+    async fn get_token(&mut self) -> Result<Token> {
+        todo!()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::token::test::MockTokenProvider;
+    use axum::response::IntoResponse;
+    use reqwest::StatusCode;
+    use serde_json::Value;
+    use tokio::task::JoinHandle;
+
+    #[test]
+    fn metadata_root_default() {
+        env::remove_var("GCE_METADATA_ROOT"); // Ensure default is used
+        env::remove_var("GCE_METADATA_HOST");
```

Comment:
`std::env::remove_var` is `unsafe`:

https://doc.rust-lang.org/std/env/fn.remove_var.html

You need to make sure the test is serialized, and we should call those functions form `unsafe` blocks. For inspiration see:

https://github.com/googleapis/google-cloud-rust/blob/30257cf2dc4ffa12a4dc5015457745f890534a9d/src/gax/src/http_client/mod.rs#L162-L169 


PS: I think we must be missing a trick in your CI if we are allowing unsafe functions to get called.

Reply:
Aha, these functions are not marked unsafe until the 2024 edition of rust.  They **are** unsafe already, just the marker is missing.  That explains why the code compiled.